### PR TITLE
executor: resolve errors caused by `in null` in point/batch ...

### LINKED ISSUE #18839: IN(NULL) causes "cannot convert <nil>(type <nil>) to string"
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
create table t ( id bigint(10) primary key, f varchar(191) default null, unique key `idx_f` (`f`));
select * from t where f in (null);
```

### 2. What did you expect to see? (Required)
This query should return no results, but it should complete without any error.

### 3. What did you see instead (Required)

```
ERROR 1105 (HY000): cannot convert <nil>(type <nil>) to string
```

### 4. Affected version (Required)

All 4.0 releases are affected.

TiDB v3.0.16 is not affected.

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #18839 

Problem Summary: resolve errors caused by `in null` in point/batch get operators

### What is changed and how it works?
Skip null values when casting.

### Check List 

Tests 

- Unit test
### Release note 

- executor: resolve errors caused by `in null` in point/batch get operators

### CODE REVIEW

**Comment by tiancaiamao** [comment]
LGTM

File: executor/point_get.go [hunk 1]
```diff
@@ -302,6 +302,9 @@ func (e *PointGetExecutor) get(ctx context.Context, key kv.Key) ([]byte, error)
 func encodeIndexKey(e *baseExecutor, tblInfo *model.TableInfo, idxInfo *model.IndexInfo, idxVals []types.Datum, tID int64) (_ []byte, err error) {
 	sc := e.ctx.GetSessionVars().StmtCtx
 	for i := range idxVals {
+		if idxVals[i].IsNull() {
```
Comment (coocood) [inline executor/point_get.go:305]: If we skip the null value, we built a non exists index key, I think it is not the right way to fix the bug.

Comment (qw4990) [reply executor/point_get.go:305]: Why does the index key not exist? This loop is used to cast the `idxVals` to target types, which is unnecessary for null type values.
After casting, all null values will also be encoded into the key.

Comment (coocood) [reply executor/point_get.go:305]: But if the index has a null value, it is not encoded as unique index, the rowid is appended to the key.

Comment (coocood) [reply executor/point_get.go:305]: I mean we should let the caller know the index value has a null value, we can not point get this key.

Comment (qw4990) [reply executor/point_get.go:305]: okay

Comment (qw4990) [reply executor/point_get.go:309]: Updated. PTAL @coocood

File: executor/point_get.go [hunk 2]
```diff
@@ -313,19 +320,19 @@ func encodeIndexKey(e *baseExecutor, tblInfo *model.TableInfo, idxInfo *model.In
 		} else {
 			idxVals[i], err = table.CastValue(e.ctx, idxVals[i], colInfo, true, false)
 			if types.ErrOverflow.Equal(err) {
-				return nil, kv.ErrNotExist
+				return nil, false, kv.ErrNotExist
```
Comment (coocood) [inline executor/point_get.go:323]: We could return `nil, true, nil`, then outer do not need to check `kv.ErrNotExist`.

Comment (qw4990) [reply executor/point_get.go:324]: After discussing with @tiancaiamao , I change the argument `nonExist` to `hasNull`. PTAL

**Comment by coocood** [comment]
LGTM

**Pull request merged**